pkg/manager/member/dm_worker_scaler.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/label"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	apps "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog"
+)
+
+type workerScaler struct {
+	generalScaler
+}
+
+// NewWorkerScaler returns a DMScaler
+func NewWorkerScaler(pvcLister corelisters.PersistentVolumeClaimLister,
+	pvcControl controller.PVCControlInterface) Scaler {
+	return &workerScaler{
+		generalScaler: generalScaler{
+			pvcLister:  pvcLister,
+			pvcControl: pvcControl,
+		},
+	}
+}
+
+func (wsd *workerScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	scaling, _, _, _ := scaleOne(oldSet, newSet)
+	if scaling > 0 {
+		return wsd.ScaleOut(meta, oldSet, newSet)
+	} else if scaling < 0 {
+		return wsd.ScaleIn(meta, oldSet, newSet)
+	}
+	return wsd.SyncAutoScalerAnn(meta, oldSet)
+}
+
+func (wsd *workerScaler) ScaleOut(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+	dc, ok := meta.(*v1alpha1.DMCluster)
+	if !ok {
+		return nil
+	}
+
+	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
+	resetReplicas(newSet, oldSet)
+	ns := dc.GetNamespace()
+	dcName := dc.GetName()
+
+	klog.Infof("scaling out dm-worker statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
+	_, err := wsd.deleteDeferDeletingPVC(dc, oldSet.GetName(), v1alpha1.DMWorkerMemberType, ordinal)
+	if err != nil {
+		return err
+	}
+
+	if !dc.Status.Worker.Synced {
+		return fmt.Errorf("DMCluster: %s/%s's dm-worker status sync failed, can't scale out now", ns, dcName)
+	}
+
+	setReplicasAndDeleteSlots(newSet, replicas, deleteSlots)
+	return nil
+}
+
+// We need remove member from cluster before reducing statefulset replicas
```

Comment:
where is the member removed?

Reply:
It will be removed in syncing worker status. For `dm-worker` we can't remove its register info from dm-master when it's still alive. So we delete it later after it's keepalive lease is outdated or revoked. We can defer deleting dm-worker register info because dm-master will patch replication worker through keepalive info.

Reply:
I mean the comment here does not match the code logic.